javascript/src/applier.ts

```diff
@@ -0,0 +1,402 @@
+  // https://html.spec.whatwg.org/multipage/rendering.html#form-controls
+  INPUT: DomAction.Skip,
+  SELECT: DomAction.Skip,
+  BUTTON: DomAction.Skip,
+  TEXTAREA: DomAction.Skip,
+
+  // Other elements where the phrase-based line breaking should be disabled.
+  // https://github.com/google/budoux/blob/main/budoux/skip_nodes.json
+  ABBR: DomAction.Skip,
+  CODE: DomAction.Skip,
+  IFRAME: DomAction.Skip,
+  TIME: DomAction.Skip,
+  VAR: DomAction.Skip,
+};
+
+/**
+ * Determine the action for an element.
+ * @param element An element to determine the action for.
+ * @returns The {@link domActions} for the element.
+ */
+function actionForElement(element: Element): DomAction {
+  const action = domActions[element.nodeName];
+  if (action !== undefined) return action;
+
+  // jsdom does not have `getComputedStyle`.
+  if (typeof getComputedStyle === 'function') {
+    const style = getComputedStyle(element);
+    switch (style.whiteSpace) {
+      case 'nowrap':
+      case 'pre':
+        return DomAction.Skip;
+    }
+
+    const display = style.display;
+    assert(display);
+    if (display !== 'inline') return DomAction.Block;
+  }
+  return DomAction.Inline;
+}
+
+/**
+ * Represents a "paragraph", broken by block boundaries or forced breaks.
+ *
+ * A CSS
+ * {@link https://drafts.csswg.org/css2/#inline-formatting inline formatting context}
+ * is usually a "paragraph", but it can be broken into multiple paragraphs by
+ * forced breaks such as `<br>`.
+ */
+class Paragraph {
+  element: HTMLElement;
+  textNodes: Text[] = [];
+
+  constructor(element: HTMLElement) {
+    this.element = element;
+  }
+
+  hasText(): boolean {
+    return this.textNodes.length > 0;
+  }
+}
+
+/**
+ * Applies the BudouX to the given DOM.
+ *
+ * This class has following advantages over
+ * {@link Applier.applyToElement}.
+ * * It recognizes paragraphs and applies the BudouX for each
+ *   paragraph separately.
+ * * It can customize how to insert break opportunities.
+ *   See {@link separator} for more details.
+ * * It is generally faster and more memory efficient, but the
+ *   code size is larger.
+ */
+export class Applier {
+  private parser_: Parser;
+  /**
+   * The threshold score to control the granularity of chunks.
+   */
+  threshold: number = DEFAULT_THRES;
+  /**
+   * The separator to insert at each semantics boundary.
+   * The default value is U+200B ZERO WIDTH SPACE.
+   *
+   * When falsy, a `<wbr>` element is inserted.
+   */
+  separator?: string = ZWSP;
+  /**
+   * This class name is added to the containing block
+   * when the BudouX is applied.
+   *
+   * The caller is responsible for defining the class.
+   * {@link defineClassAs} can append a `<style>` element
+   * that defines the default styles as a class.
+   *
+   * When falsy, an inline style is set instead.
+   */
+  className?: string;
+
+  /**
+   * @param parser A BudouX {@link Parser} to compute semantic line breaks.
+   */
+  constructor(parser: Parser) {
```

Comment:
Can we add an `options` arg to make those properties (e.g. `separator`) configurable when instantiating one?

Reply:
Good idea, thanks, done.